### PR TITLE
Ixxat bus state and hardware errors detection

### PR DESCRIPTION
**Pull request by cowo78**
Implemented IXXATBus.state()  property.
Hardware errors are detected in _recv_internal() and raise an exception.

### CODE REVIEW

**Comment by felixdivo** [comment]
#1119 created some conflicts

**Comment by felixdivo** [comment]
@cowo78 I can do a review once they are resolved.

File: can/interfaces/ixxat/canlib.py [hunk 1]
```diff
@@ -140,8 +143,18 @@ def _recv_internal(self, timeout):
     def send(self, msg: Message, timeout: Optional[float] = None) -> None:
         return self.bus.send(msg, timeout)
 
-    def _send_periodic_internal(self, msg, period, duration=None):
+    def _send_periodic_internal(self, msgs, period, duration=None):
```
Comment (felixdivo) [inline can/interfaces/ixxat/canlib.py:146]: Good change! Please also rename `msg` in the next line and the two receivers, to really restore consistence here

File: can/interfaces/ixxat/canlib_vcinpl.py [hunk 2]
```diff
@@ -639,85 +648,92 @@ def flush_tx_buffer(self):
...
+            # See if we got a data or info/error messages
+            if self._message.uMsgInfo.Bits.type == constants.CAN_MSGTYPE_DATA:
+                data_received = True
+            elif self._message.uMsgInfo.Bits.type == constants.CAN_MSGTYPE_INFO:
+                log.info(
+                    CAN_INFO_MESSAGES.get(
+                        self._message.abData[0],
+                        "Unknown CAN info message code {}".format(
+                            self._message.abData[0]
+                        ),
```
Comment (felixdivo) [inline can/interfaces/ixxat/canlib_vcinpl.py:689]:
```suggestion
                        f"Unknown CAN info message code {self._message.abData[0]}",
```

File: can/interfaces/ixxat/canlib_vcinpl.py [hunk 3]
```diff
@@ -639,85 +648,92 @@ def flush_tx_buffer(self):
...
-                                "Unknown CAN error message code {}".format(
-                                    self._message.abData[0]
-                                ),
-                            )
+                    log.warning(
+                        CAN_ERROR_MESSAGES.get(
+                            self._message.abData[0],
+                            "Unknown CAN error message code {}".format(
+                                self._message.abData[0]
+                            ),
```
Comment (felixdivo) [inline can/interfaces/ixxat/canlib_vcinpl.py:701]:
```suggestion
                            f"Unknown CAN error message code {self._message.abData[0]}",
```

File: can/interfaces/kvaser/canlib.py [hunk 4]
```diff
@@ -330,6 +345,16 @@ def init_kvaser_library():
...
+def init_kvaser_library():
+    if __canlib is not None:
+        try:
+            log.debug("Initializing Kvaser CAN library")
+            canInitializeLibrary()
+            log.debug("CAN library initialized")
+        except:
+            log.warning("Kvaser canlib could not be initialized.")
+
+
```
Comment (felixdivo) [inline can/interfaces/kvaser/canlib.py:357]: Isn't this a duplicate?

File: can/interfaces/kvaser/canlib.py [hunk 5]
```diff
@@ -319,6 +319,21 @@ def __check_bus_handle_validity(handle, function, arguments):
...
+    )
+
+    if sys.platform == "win32":
+        canGetVersionEx = __get_canlib_function(
+            "canGetVersionEx",
+            argtypes=[ctypes.c_uint],
+            restype=ctypes.c_uint,
+            errcheck=__check_status,
+        )
+
```
Comment (felixdivo) [inline can/interfaces/kvaser/canlib.py:336]: In what way are these changes related to IXXAT?

**Review by felixdivo — CHANGES_REQUESTED** [review]
Thank you for the contribution.

Is is possible that `can/interfaces/ixxat/canlib_vcinpl2.py` would need to be changed too?

File: can/interfaces/kvaser/canlib.py (hunk 4 above)
Comment (cowo78) [reply can/interfaces/kvaser/canlib.py]: My bad, I reverted commit ad11bf95 that introduced the (unwanted) change.

File: can/interfaces/kvaser/canlib.py (hunk 5 above)
Comment (cowo78) [reply can/interfaces/kvaser/canlib.py]: My bad, I reverted commit ad11bf95 that introduced the (unwanted) change.

**Comment by cowo78** [comment]
> Is is possible that `can/interfaces/ixxat/canlib_vcinpl2.py` would need to be changed too?

Changed the f-strings in vcinpl2 too.

**Review by felixdivo — APPROVED** [review]

**Comment by felixdivo** [comment]
Am I right in that this only adds `bus.state` support for `can/interfaces/ixxat/canlib_vcinpl1.py` and not for `can/interfaces/ixxat/canlib_vcinpl2.py`?

**Comment by cowo78** [comment]
> Am I right in that this only adds `bus.state` support for `can/interfaces/ixxat/canlib_vcinpl1.py` and not for `can/interfaces/ixxat/canlib_vcinpl2.py`?

Correct. As mentioned in #1119 I would suggest to postpone a refactoring / code deduplication do another PR.

**Comment by felixdivo** [comment]
I'd like this to get a single other review before merging.

**Comment by felixdivo** [comment]
@hardbyte? 😃

**Comment by zariiii9003** [comment]
@cowo78, @fjburgos any objections to merging this?

**Review by hardbyte — APPROVED** [review]
I've read through this patch while I don't have any issues with it I don't have the hardware to test it.
